hackathon_site/dashboard/frontend/src/slices/hardware/hardwareSlice.ts

```diff
@@ -1,48 +1,150 @@
+
+        try {
+            const response = await get<APIListResponse<Hardware>>(
+                "/api/hardware/hardware/",
+                filters
+            );
+            return response.data;
+        } catch (e: any) {
+            dispatch(
+                displaySnackbar({
+                    message: `Failed to fetch hardware data: Error ${e.response.status}`,
+                    options: { variant: "error" },
+                })
+            );
+            return rejectWithValue({
+                status: e.response.status,
+                message: e.response.data,
+            });
+        }
+    }
+);
+
+// Slice
+
 const hardwareSlice = createSlice({
     name: hardwareReducerName,
-    initialState: hardwareAdapter.getInitialState(extraState),
-    reducers: {},
+    initialState,
+    reducers: {
+        /**
+         * Update the filters for the Hardware API
+         *
+         * To clear a particular filter, set the field to undefined.
+         */
+        setFilters: (
+            state: HardwareState,
+            { payload }: PayloadAction<HardwareFilters>
+        ) => {
+            const { in_stock, hardware_ids, category_ids, search, ordering } = {
+                ...state.filters,
+                ...payload,
+            };
+
+            // Remove values that are empty or falsy
+            state.filters = {
+                ...(in_stock && { in_stock }),
+                ...(hardware_ids && hardware_ids.length > 0 && { hardware_ids }),
+                ...(category_ids && category_ids.length > 0 && { category_ids }),
+                ...(search && { search }),
+                ...(ordering && { ordering }),
+            };
+        },
+
+        clearFilters: (
+            state: HardwareState,
+            { payload }: PayloadAction<{ saveSearch?: boolean } | undefined>
+        ) => {
+            const { search } = state.filters;
+
+            state.filters = {};
+
+            if (payload?.saveSearch) {
+                state.filters.search = search;
+            }
+        },
+    },
+    extraReducers: (builder) => {
+        builder.addCase(getHardwareWithFilters.pending, (state) => {
+            state.isLoading = true;
+            state.error = null;
+        });
+
+        builder.addCase(getHardwareWithFilters.fulfilled, (state, { payload }) => {
+            state.isLoading = false;
+            state.error = null;
+            state.next = payload.next;
+            hardwareAdapter.setAll(state, payload.results);
+        });
+
+        builder.addCase(getHardwareWithFilters.rejected, (state, { payload }) => {
+            state.isLoading = false;
+            state.error = payload?.message || "Something went wrong";
+        });
+    },
 });
 
 export const { actions, reducer } = hardwareSlice;
 export default reducer;
 
+export const { setFilters, clearFilters } = actions;
+
 // Selectors
 export const hardwareSliceSelector = (state: RootState) => state[hardwareReducerName];
```

Comment:
So first, a state is pretty much just a javascript object.

A reducer is a function that takes in a state and an action, and returns an updated version of the state (reducers are a generic concept, they aren't redux-specific). In redux toolkit land, "Reducers" are a bit more all-encompasing, since it does also contain definitions for its state and actions - but that is a very Redux Toolkit thing.

And like you said, a selector takes in a state and returns something in that state (could be a specific field, could be an object within that state that has more fields).

Since `state: RootState` is basically just an object, all this function is doing is returning the slice of that object at `hardwareReducerName`. 

How that state object is created is wrapped up in redux abstractions for us. In `store.ts`, we define a map of reducer names to reducers (Redux Toolkiti reducers, so wrappers for the actual reducers and functions); `configureStore()` then takes that object, and builds the initial state from it. How it actually does that I'm not sure, but you could dig into the redux toolkit source code to figure it out!